Mathlib/Topology/Order/Basic.lean

```diff
@@ -0,0 +1,2795 @@
+    ∃ᶠ x in 𝓝 a, x ∈ s :=
+  (ha.frequently_mem hs).filter_mono inf_le_left
+#align is_glb.frequently_nhds_mem IsGLB.frequently_nhds_mem
+
+theorem IsLUB.mem_closure {a : α} {s : Set α} (ha : IsLUB s a) (hs : s.Nonempty) : a ∈ closure s :=
+  (ha.frequently_nhds_mem hs).mem_closure
+#align is_lub.mem_closure IsLUB.mem_closure
+
+theorem IsGLB.mem_closure {a : α} {s : Set α} (ha : IsGLB s a) (hs : s.Nonempty) : a ∈ closure s :=
+  (ha.frequently_nhds_mem hs).mem_closure
+#align is_glb.mem_closure IsGLB.mem_closure
+
+theorem IsLUB.nhdsWithin_neBot {a : α} {s : Set α} (ha : IsLUB s a) (hs : s.Nonempty) :
+    NeBot (𝓝[s] a) :=
+  mem_closure_iff_nhdsWithin_neBot.1 (ha.mem_closure hs)
+#align is_lub.nhds_within_ne_bot IsLUB.nhdsWithin_neBot
+
+theorem IsGLB.nhdsWithin_neBot : ∀ {a : α} {s : Set α}, IsGLB s a → s.Nonempty → NeBot (𝓝[s] a) :=
+  @IsLUB.nhdsWithin_neBot αᵒᵈ _ _ _
+#align is_glb.nhds_within_ne_bot IsGLB.nhdsWithin_neBot
+
+theorem isLUB_of_mem_nhds {s : Set α} {a : α} {f : Filter α} (hsa : a ∈ upperBounds s) (hsf : s ∈ f)
+    [NeBot (f ⊓ 𝓝 a)] : IsLUB s a :=
+  ⟨hsa, fun b hb =>
+    not_lt.1 fun hba =>
+      have : s ∩ { a | b < a } ∈ f ⊓ 𝓝 a := inter_mem_inf hsf (IsOpen.mem_nhds (isOpen_lt' _) hba)
+      let ⟨_x, ⟨hxs, hxb⟩⟩ := Filter.nonempty_of_mem this
+      have : b < b := lt_of_lt_of_le hxb <| hb hxs
+      lt_irrefl b this⟩
+#align is_lub_of_mem_nhds isLUB_of_mem_nhds
+
+theorem isLUB_of_mem_closure {s : Set α} {a : α} (hsa : a ∈ upperBounds s) (hsf : a ∈ closure s) :
+    IsLUB s a := by
+  rw [mem_closure_iff_clusterPt, ClusterPt, inf_comm] at hsf
+  exact isLUB_of_mem_nhds hsa (mem_principal_self s)
+#align is_lub_of_mem_closure isLUB_of_mem_closure
+
+theorem isGLB_of_mem_nhds :
+    ∀ {s : Set α} {a : α} {f : Filter α}, a ∈ lowerBounds s → s ∈ f → NeBot (f ⊓ 𝓝 a) → IsGLB s a :=
+  @isLUB_of_mem_nhds αᵒᵈ _ _ _
+#align is_glb_of_mem_nhds isGLB_of_mem_nhds
+
+theorem isGLB_of_mem_closure {s : Set α} {a : α} (hsa : a ∈ lowerBounds s) (hsf : a ∈ closure s) :
+    IsGLB s a :=
+  @isLUB_of_mem_closure αᵒᵈ _ _ _ s a hsa hsf
+#align is_glb_of_mem_closure isGLB_of_mem_closure
+
+theorem IsLUB.mem_upperBounds_of_tendsto [Preorder γ] [TopologicalSpace γ] [OrderClosedTopology γ]
+    {f : α → γ} {s : Set α} {a : α} {b : γ} (hf : MonotoneOn f s) (ha : IsLUB s a)
+    (hb : Tendsto f (𝓝[s] a) (𝓝 b)) : b ∈ upperBounds (f '' s) := by
+  rintro _ ⟨x, hx, rfl⟩
+  replace ha := ha.inter_Ici_of_mem hx
+  haveI := ha.nhdsWithin_neBot ⟨x, hx, le_rfl⟩
+  refine' ge_of_tendsto (hb.mono_left (nhdsWithin_mono _ (inter_subset_left s (Ici x)))) _
+  exact mem_of_superset self_mem_nhdsWithin fun y hy => hf hx hy.1 hy.2
+#align is_lub.mem_upper_bounds_of_tendsto IsLUB.mem_upperBounds_of_tendsto
+
+-- For a version of this theorem in which the convergence considered on the domain `α` is as `x : α`
+-- tends to infinity, rather than tending to a point `x` in `α`, see `is_lub_of_tendsto_atTop`
```

Comment:
```suggestion
-- For a version of this theorem in which the convergence considered on the domain `α` is as `x : α`
-- tends to infinity, rather than tending to a point `x` in `α`, see `isLUB_of_tendsto_atTop`
```